src/Control/Monad/Logic/Sequence/Internal.hs

```diff
@@ -174,16 +175,27 @@ type Seq = SeqT Identity
 
 fromView :: m (View m a) -> SeqT m a
 fromView = SeqT . S.singleton
-{-# INLINE fromView #-}
+{-# INLINE [1] fromView #-}
 
 toView :: Monad m => SeqT m a -> m (View m a)
 toView (SeqT s) = case S.viewl s of
   S.EmptyL -> return Empty
   h S.:< t -> h >>= \x -> case x of
     Empty -> toView (SeqT t)
     hi :< SeqT ti -> return (hi :< SeqT (ti S.>< t))
-{-# INLINEABLE toView #-}
-{-# SPECIALIZE INLINE toView :: Seq a -> Identity (View Identity a) #-}
+{-# INLINEABLE [1] toView #-}
+
+-- For now, we don't assume the monad identity law holds for the underlying
+-- monad. We may re-evaluate that later, but it's a bit tricky to document the
+-- detailed strictness requirements properly.
+--
+-- We do, however, assume that `pure /= _|_`, or that `>>=` doesn't `seq` on
+-- its second argument, and that we can therefore eta-reduce `\x -> pure x` to
+-- just `pure`. It seems quite safe to assume that at least one of these is
+-- true, since in real code they're virtually always *both* true.
+{-# RULES
+"toView . fromView" forall m. toView (fromView m) = m >>= return
```

Comment:
Do you see any performance difference with this? I guess what I'm actually wondering is: does this ever fire?